### PR TITLE
upgrade to spring boot2

### PR DESCRIPTION
**Pull request by jdfox**
# Motivation and Context

Spring Boot 1.x had its end-of-life in August 2019.
Maintenance for Spring Boot 2.1 will end 1 November 2020, so it's time to upgrade Spring Boot to 2.2.
Spring Boot 2 also brings some useful changes, including a welcome simplification of security settings.

# What has changed
- Updated Spring Framework to 5.2, and Spring Boot to 2.2.
- Updated lib dependencies, migrated deprecated config in YAML, with security config in YAML replaced by a new class SecurityConfig.java.
- copied in missing `ctp.common.error.CTPRetryPolicy` from Action Service, and updated its imports for SB 2.x
- replaced godaddy-logger with slf4j

# How to test?
- `mvn clean install`
- `mvn spring-boot:run`
- deploy gateway with new image to a sandbox cluster
- set bucket in sandbox to a test bucket using ras-rm-cli
- run acceptance tests on sandbox cluster 
- use port forwarding to connect to Response Ops on localhost:8085, and create a new collection exercise for same day
- check that the print file is in the test bucket 

# Links
- https://collaborate2.ons.gov.uk/jira/browse/RAS-82

### CODE REVIEW

File: src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java [hunk 1]
```diff
@@ -31,9 +31,13 @@
    */
   @ExceptionHandler(CTPException.class)
   public ResponseEntity<?> handleCTPException(CTPException exception) {
-    log.with("fault", exception.getFault())
-        .with("exception_message", exception.getMessage())
-        .error("Uncaught CTPException", exception);
+    log.error(
```
Comment (insacuri) [inline src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java:34]: Does this logger not do structured logging?

File: _infra/helm/action-exporter/Chart.yaml [hunk 2]
```diff
@@ -18,5 +18,5 @@ version: 1.2.0
 
 # This is the version number of the application being deployed. This version number should be
 # incremented each time you make changes to the application.
-appVersion: 12.2.0
+appVersion: 12.2.1
```
Comment (warrenbailey) [inline _infra/helm/action-exporter/Chart.yaml:21]: I think we should increment either the major or minor version, this is quite a big upgrade. Maybe 12.3.1 ?

File: pom.xml [hunk 3]
```diff
@@ -10,13 +10,12 @@
     <description>CTP : ActionExporterService</description>
 
     <properties>
-        <springboot.version>1.5.6.RELEASE</springboot.version>
-        <springcloud.version>Dalston.SR1</springcloud.version>
+        <springboot.version>2.2.9.RELEASE</springboot.version>
+        <springcloud.version>Hoxton.SR7</springcloud.version>
         <common.version>11.0.0-SNAPSHOT</common.version>
```
Comment (warrenbailey) [inline pom.xml:15]: Is this common.version correct? seems odd to be SNAPSHOT?

File: pom.xml [hunk 4]
```diff
@@ -231,6 +234,31 @@
             <scope>compile</scope>
         </dependency>
 
+        <dependency>
+            <groupId>net.bytebuddy</groupId>
+            <artifactId>byte-buddy</artifactId>
+            <version>1.10.13</version>
```
Comment (warrenbailey) [inline pom.xml:240]: We should paramaterize these versions with the ones above.

File: pom.xml [hunk 5]
```diff
@@ -231,6 +234,31 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>net.bytebuddy</groupId>
+            <artifactId>byte-buddy-dep</artifactId>
+            <version>1.10.13</version>
```
Comment (warrenbailey) [inline pom.xml:246]: as above

File: pom.xml [hunk 6]
```diff
@@ -231,6 +234,31 @@
...
+        </dependency>
+
+        <dependency>
+            <groupId>javax.xml.soap</groupId>
+            <artifactId>javax.xml.soap-api</artifactId>
+            <version>1.4.0</version>
```
Comment (warrenbailey) [inline pom.xml:252]: as above

File: pom.xml [hunk 7]
```diff
@@ -231,6 +234,31 @@
...
+        </dependency> 
+
+        <dependency>
+            <groupId>com.sun.xml.messaging.saaj</groupId>
+            <artifactId>saaj-impl</artifactId>
+            <version>1.5.2</version>
```
Comment (warrenbailey) [inline pom.xml:258]: as above

File: src/main/java/uk/gov/ons/ctp/common/retry/CTPRetryPolicy.java [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+public class CTPRetryPolicy implements RetryPolicy {
+
+  private static final Logger log = LoggerFactory.getLogger(CTPRetryPolicy.class);
+
+  private static final int DEFAULT_MAX_ATTEMPTS = 3;
+  private static final String RUNTIME_EXCEPTION = "java.lang.RuntimeException";
+
+  private volatile int maxAttempts;
+  private volatile List<String>
+      retryableExceptions; // TODO Make it a List<Class<? extends Throwable>
```
Comment (warrenbailey) [inline src/main/java/uk/gov/ons/ctp/common/retry/CTPRetryPolicy.java:31]: Weird formatting here, could it just be on one line? Can the TODO go?

Why do these need to be volatile?

File: src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java (hunk 1 above)
Comment (warrenbailey) [reply src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java:34]: I don't think it does out of the box, you need to use something like logback. Which might be worth doing tbh.

File: _infra/helm/action-exporter/Chart.yaml (hunk 2 above)
Comment (jdfox) [reply _infra/helm/action-exporter/Chart.yaml:21]: Agreed.

File: pom.xml (hunk 3 above)
Comment (jdfox) [reply pom.xml]: Not only that, but it's currently unused. I've now removed it.

File: pom.xml (hunk 4 above)
Comment (jdfox) [reply pom.xml:239]: done

File: pom.xml (hunk 5 above)
Comment (jdfox) [reply pom.xml:245]: done

File: pom.xml (hunk 6 above)
Comment (jdfox) [reply pom.xml]: done

File: pom.xml (hunk 7 above)
Comment (jdfox) [reply pom.xml]: unpinned

File: src/main/java/uk/gov/ons/ctp/common/retry/CTPRetryPolicy.java (hunk 8 above)
Comment (jdfox) [reply src/main/java/uk/gov/ons/ctp/common/retry/CTPRetryPolicy.java]: Formatting fixed, TODO removed. 

The vars might not need to be volatile: this RetryPolicy (called from springintegration/action-instruction-inbound-flow.xml) was missing, so I used the one from Action service.  There's no harm in leaving the vars as volatile but, especially in a container, they might not need to be.

File: src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java (hunk 1 above)
Comment (jdfox) [reply src/main/java/uk/gov/ons/ctp/common/error/RestExceptionHandler.java:34]: Logback is already installed, but unused. Do we need these errors formatted for Stackdriver?

**Review by warrenbailey — APPROVED** [review]

**Comment by jdfox** [comment]
Branch restored pending test and possible patch of previous change.

**Pull request merged**